ledger/cryptarchia-ledger/src/lib.rs

```diff
@@ -69,22 +70,32 @@ impl EpochState {
 }
 
 #[derive(Clone, Debug, PartialEq)]
-pub struct Ledger {
-    states: HashMap<HeaderId, LedgerState>,
+pub struct Ledger<Id: Eq + Hash> {
+    states: HashMap<Id, LedgerState>,
     config: Config,
 }
 
-impl Ledger {
-    pub fn from_genesis(id: HeaderId, state: LedgerState, config: Config) -> Self {
+impl<Id> Ledger<Id>
+where
+    Id: Eq + Hash + Copy,
+{
+    pub fn from_genesis(id: Id, state: LedgerState, config: Config) -> Self {
         Self {
             states: [(id, state)].into_iter().collect(),
             config,
         }
     }
 
     #[must_use = "this returns the result of the operation, without modifying the original"]
-    pub fn try_apply_header(&self, header: &Header) -> Result<Self, LedgerError> {
-        let parent_id = header.parent();
+    pub fn try_update(
+        &self,
+        id: Id,
+        parent_id: Id,
+        slot: Slot,
+        proof: &LeaderProof,
+        // (update corresponding to the leader proof, leader proof)
+        orphan_proofs: impl Iterator<Item = (Id, LeaderProof)>,
```

Comment:
I prefer `IntoIterator`, it lets you pass in arbitrary collections without first converting to iterator
```suggestion
        orphan_proofs: impl IntoIterator<Item = (Id, LeaderProof)>,
```